tests/PySys/software-management-end-to-end/environment_sm_management.py

```diff
@@ -0,0 +1,267 @@
+        So far, no checks are done on the json_content.
+        """
+
+        url = f"https://{self.tenant_url}.cumulocity.com/devicecontrol/operations"
+
+        payload = {
+            "deviceId": self.project.deviceid,
+            "description": "Apply software changes, triggered from PySys test",
+            "c8y_SoftwareUpdate": json_content,
+        }
+
+        req = requests.post(url, json=payload, headers=self.header)
+
+        jresponse = json.loads(req.text)
+
+        self.log.info("Response status: %s", req.status_code)
+        self.log.info("Response to action: %s", json.dumps(jresponse, indent=4))
+
+        self.operation = jresponse
+        self.operation_id = jresponse.get("id")
+
+        if not self.operation_id:
+            raise SystemError("field id is mising in response")
+
+        self.log.info("Started operation: %s", self.operation)
+
+        req.raise_for_status()
+
+    def is_status_fail(self):
+        """Check if the current status is a fail"""
+        if self.operation_id:
+            return self.check_status_of_operation("FAILED")
+        return self.check_last_status("FAILED")
+
+    def is_status_success(self):
+        """Check if the current status is a success"""
+        if self.operation_id:
+            return self.check_status_of_operation("SUCCESSFUL")
+        return self.check_last_status("SUCCESSFUL")
+
+    def check_last_status(self, status):
+        """Check if the last operation is successfull.
+        Warning: an observation so far is, that installation failures
+        seem to be at the beginning of the list independent of if we
+        revert it or not.
+        """
+
+        params = {
+            "deviceId": self.project.deviceid,
+            "pageSize": 1,
+            # To get the latest records first
+            "revert": "true",
+            # By using the date we make sure that the request comes
+            # sorted, otherwise the revert does not seem to have an
+            # effect. The lower boundary seems to be ok so we just
+            # use the beginning of the epoch same as the c8y ui.
+            "dateFrom": "1970-01-01T00:00:00.000Z",
+        }
+
+        url = f"https://{self.tenant_url}.cumulocity.com/devicecontrol/operations"
+        req = requests.get(url, params=params, headers=self.header)
+
+        req.raise_for_status()
+
+        self.log.debug("Final URL of the request: %s", req.url)
+
+        jresponse = json.loads(req.text)
+
+        if not jresponse["operations"]:
+            self.log.error("No operations found")
+            return None
+
+        # Get the last operation, when we set "revert": "true" we can read it
+        # from the beginning of the list
+
+        operations = jresponse.get("operations")
+
+        if not operations or len(operations)!=1:
+            raise SystemError("field operations is mising in response or to long")
+
+        operation = operations[0]
```

Comment:
Why are we relying on the order and not doing the lookup based on the operation `id` even if there are multiple operations?

Reply:
When we already have an operation id then we can use method check_status_of_operation. As long as we do not expect multiple running operations here, it would be easier just to get the first one.